File: PHP/0008.php
class Solution {

/**
 * @param String $s
 * @return Integer
 */
function myAtoi($s) {
    // Define the 32-bit signed integer boundaries
    $INT_MIN = -2147483648;
    $INT_MAX = 2147483647;
    
    // Trim leading whitespace
    $s = trim($s);
    if (strlen($s) === 0) {
        return 0;
    }
    
    // Initialize the sign and result
    $sign = 1;
    $index = 0;
    $result = 0;
    
    // Check for sign
    if ($s[$index] === '-') {
        $sign = -1;
        $index++;
    } else if ($s[$index] === '+') {
        $index++;
    }
    
    // Process the digits
    while ($index < strlen($s) && ctype_digit($s[$index])) {
        $digit = intval($s[$index]);
        
        // Check for overflow before updating result
        if ($result > ($INT_MAX - $digit) / 10) {
            return $sign === 1 ? $INT_MAX : $INT_MIN;
        }
        
        $result = $result * 10 + $digit;
        $index++;
    }
    
    return $sign * $result;
}
}

File: Ruby/0008.rb
# @param {String} s
# @return {Integer}
def my_atoi(s)
    # Define the 32-bit signed integer boundaries
    int_min = -2**31
    int_max = 2**31 - 1

    # Trim leading whitespace
    s = s.strip
    return 0 if s.empty?

    # Initialize sign and result
    sign = 1
    result = 0
    i = 0

    # Determine the sign
    if s[i] == '-'
        sign = -1
        i += 1
    elsif s[i] == '+'
        i += 1
    end

    # Process the digits
    while i < s.length && s[i].match?(/\d/)
        digit = s[i].to_i
        
        # Check for overflow before updating the result
        if result > (int_max - digit) / 10
            return sign == 1 ? int_max : int_min
        end
        
        result = result * 10 + digit
        i += 1
    end

    sign * result
end
